history/src/block_stores/multiple_strategy_block_store.rs

```diff
@@ -0,0 +1,136 @@
+    traits::block_storage_interface::{BlockStorageImpl, BlockStorageInterface},
+};
+use solana_rpc_client::nonblocking::rpc_client::RpcClient;
+use solana_rpc_client_api::config::RpcBlockConfig;
+use solana_sdk::{
+    commitment_config::{CommitmentConfig, CommitmentLevel},
+    slot_history::Slot,
+};
+use std::{
+    ops::Range,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+pub struct MultipleStrategyBlockStorage {
+    inmemory_for_storage: InmemoryBlockStore, // for confirmed blocks
+    persistent_block_storage: BlockStorageImpl, // for persistent block storage
+    faithful_rpc_client: Option<Arc<RpcClient>>, // to fetch legacy blocks from faithful
+    last_confirmed_slot: Arc<AtomicU64>,
+}
+
+impl MultipleStrategyBlockStorage {
+    pub fn new(
+        persistent_block_storage: BlockStorageImpl,
+        faithful_rpc_client: Option<Arc<RpcClient>>,
+        number_of_slots_in_memory: usize,
+    ) -> Self {
+        Self {
+            inmemory_for_storage: InmemoryBlockStore::new(number_of_slots_in_memory),
+            persistent_block_storage,
+            faithful_rpc_client,
+            last_confirmed_slot: Arc::new(AtomicU64::new(0)),
+        }
+    }
+
+    pub async fn get_in_memory_block(&self, slot: Slot) -> Option<ProducedBlock> {
+        self.inmemory_for_storage.get(slot, RpcBlockConfig { encoding: None, transaction_details: None, rewards: None, commitment: None, max_supported_transaction_version: None}).await
+    }
+}
+
+#[async_trait]
+impl BlockStorageInterface for MultipleStrategyBlockStorage {
+    async fn save(
+        &self,
+        slot: solana_sdk::slot_history::Slot,
+        block: ProducedBlock,
+        commitment: solana_sdk::commitment_config::CommitmentLevel,
+    ) {
+        if commitment == CommitmentLevel::Confirmed {
+            self.inmemory_for_storage
+                .save(slot, block, commitment)
+                .await;
+        } else if commitment == CommitmentLevel::Finalized {
+            let block_in_mem = self.get_in_memory_block(slot).await;
+            if block_in_mem.is_none() {
+                self.inmemory_for_storage
+                    .save(slot, block.clone(), commitment)
+                    .await;
+            } else if let Some(block_in_mem) = block_in_mem {
+                // check if inmemory blockhash is same as finalized, update it if they are not
+                // we can have two machines with same identity publishing two different blocks on same slot
+                if block_in_mem.blockhash != block.blockhash {
+                    self.inmemory_for_storage
+                        .save(slot, block.clone(), commitment)
+                        .await;
+                }
+            }
+            self.persistent_block_storage
+                .save(slot, block, commitment)
+                .await;
+        }
+        if slot > self.last_confirmed_slot.load(Ordering::Relaxed) {
```

Comment:
could use compare_exchange

Reply:
compare exchange requires anyways to fetch the value and the value will be exchanged only if it is the same as current value which may add additional complexity.